### PR TITLE
feat(python): expand Map API to cover more Add Data types (#246)

### LINKED ISSUE #246: Improve the Python API to support more data types
All data types supported by the Add Data menu should be supported by the Python API so that users can add data programmatically

### PR DESCRIPTION
**Pull request by giswqs**
## Summary

Closes #246. Expands the `geolibre` Python widget API so users can add — programmatically — the URL/cloud-native data types the desktop **Add Data** menu supports, going well beyond the previous `add_geojson` / `add_tile_layer` / `add_cog`.

### New `Map` methods

| Category | Methods |
| --- | --- |
| Web services | `add_wms`, `add_wmts`, `add_wfs` |
| Vector files | `add_vector`, `add_geoparquet`, `add_flatgeobuf`, `add_shp` |
| Tiled / cloud | `add_vector_tiles`, `add_pmtiles` |
| 3D / media | `add_3d_tiles`, `add_video` |
| Raster | `add_raster` (alias of `add_cog`) |

### How it works

Each method builds a project-layer record in `project.py` that mirrors the corresponding TypeScript store-layer shape (`createVectorStoreLayer`, `createWmsTileUrl`, the PMTiles/3D-Tiles/video records, etc.). Because every Python `add_*` call posts a new project → `loadProject` bumps `projectGeneration` → the `DesktopShell` restore effect re-runs idempotently, the embedded app reconstructs these layers through the **same restore paths and core layer sync** it uses for layers added through the UI (`restoreVectorLayers`, `restoreRasterLayers`, `restoreThreeDTilesLayers`, and `syncLayer`). No frontend changes were needed.

### Local files

`add_vector` (and the format wrappers) accept a remote URL — streamed by the in-browser vector control — or a **local** file path, which the kernel reads with GeoPandas (optional dependency) and inlines as GeoJSON, since the browser cannot read a kernel-side file. A missing GeoPandas raises a clear, actionable error.

### Out of scope (follow-ups)

- **ArcGIS** — needs a new upstream `restoreArcGISLayers` (no restore path exists today; `addArcGISLayer` runs only at UI add-time).
- **deck.gl viz** — needs full `vizConfig` + inline-data plumbing.
- **Zarr / LiDAR / Gaussian splat** — niche; each needs restore verification.
- **Local MBTiles / OSM PBF / local raster reads** — Tauri/browser-only, not feasible from the Jupyter kernel.

## Test plan

- [x] `pytest` — 60 passed (new builder-shape tests in `test_project.py`; new `Map`-method tests in `test_map.py` that stub the static server so `Map()` constructs without a bundle).
- [x] `ruff check` / `ruff format` clean.
- [x] `pre-commit run` (incl. the `npm build` hook) passes — the full app still compiles.
- [ ] Manual: load a `.geolibre.json` containing one layer of each new type in the web app and confirm it renders (recommended before release).

## Summary by CodeRabbit

* **New Features**
  * Added many new layer types: local vector files (GeoParquet, FlatGeobuf, Shapefile), vector tiles & PMTiles, OGC services (WMS, WMTS, WFS), 3D tiles, georeferenced video, and a raster alias; unified vector add helper handles URLs, geo-interface objects, and local inlining (with warnings for ignored params).

* **Documentation**
  * Expanded API reference with signatures/descriptions for all new layer helpers and clarified when optional extras are required (local vector inlining and GeoDataFrame vs remote streaming).

* **Tests**
  * Added comprehensive unit tests for all new layer helpers, inlining/WFS behavior, validations, and edge cases.

### CODE REVIEW

File: python/src/geolibre/project.py [hunk 1]
```diff
@@ -304,6 +309,413 @@ def cog_layer(
...
+    if len(coordinates) != 4 or any(len(corner) != 2 for corner in coordinates):
+        raise ValueError(
+            "coordinates must be four [lng, lat] corners (top-left, top-right, "
+            "bottom-right, bottom-left)"
+        )
+    layer = _layer_base(name, "video", **style)
+    layer["source"] = {
+        "urls": clean_urls,
+        "coordinates": [[float(c[0]), float(c[1])] for c in coordinates],
+    }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py:761]: Done in f2a34b1 — `video_layer` now sets `source.type:"video"`, `metadata.sourceKind:"video-url"`, and the corner bbox `metadata.bounds` ([west, south, east, north]) so "Zoom to layer"/`fitLayer` works like a UI-added video layer.

File: python/src/geolibre/geolibre.py [hunk 2]
```diff
@@ -288,6 +325,350 @@ def add_cog(
...
+        url = _project.wfs_getfeature_url(
+            endpoint,
+            type_name,
+            version=version,
+            output_format=output_format,
+            srs_name=srs_name,
+            max_features=max_features,
+        )
+        fc = _project.load_featurecollection(url)
+        layer = _project.geojson_layer(name, fc, source_url=url, **style)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py:482]: Done in f2a34b1 — `add_wfs` now persists the protocol fields (`service`, `typeName`, `version`, `outputFormat`, and `srsName` when set) on `source` so the Edit Layer panel can pre-populate and `isWfsLayer()` recognizes a Python-produced layer.

File: python/src/geolibre/geolibre.py [hunk 3]
```diff
@@ -288,6 +325,350 @@ def add_cog(
...
+            urls: One video URL or a list of format fallbacks (e.g. MP4, WebM).
+            coordinates: Four ``[lng, lat]`` corners in top-left, top-right,
+                bottom-right, bottom-left order.
+            name: Layer display name.
+            **style: Style overrides.
+
+        Returns:
+            The id of the added layer.
+        """
+        url_list = [urls] if isinstance(urls, str) else list(urls)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py:729]: Done in f2a34b1 — the https:// guard is now enforced in the `video_layer` builder (which `add_video` calls), so the check covers both the public method and direct builder use. Non-https URLs raise a clear ValueError up front.

File: python/src/geolibre/geolibre.py [hunk 4]
```diff
@@ -24,6 +24,43 @@
...
+        raise ImportError(
+            "Reading a local vector file requires GeoPandas. Install it with "
+            "`pip install geopandas`, or pass a URL to a hosted dataset instead."
+        ) from exc
+    gdf = geopandas.read_file(file_path)
+    if gdf.crs is not None:
+        gdf = gdf.to_crs(epsg=4326)
+    # Round-trip through GeoPandas' own GeoJSON writer so numpy/datetime property
+    # values become plain JSON the widget bus can serialize.
+    return json.loads(gdf.to_json())
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py]: Done in f2a34b1 — both addressed: (1) `_read_local_vector` now caps the converted GeoJSON at `_project._MAX_GEOJSON_BYTES` (50 MB); (2) `.parquet`/`.geoparquet`/`.pq` files dispatch to `geopandas.read_parquet` instead of the GDAL-driver-dependent `read_file`.

File: python/src/geolibre/project.py [hunk 5]
```diff
@@ -304,6 +309,413 @@ def cog_layer(
...
+        type_name: WFS feature type name (e.g. ``"topp:states"``).
+        version: WFS protocol version (e.g. ``"2.0.0"`` or ``"1.1.0"``).
+        output_format: Requested output format.
+        srs_name: Spatial reference of the response.
+        max_features: Optional cap on the number of returned features.
+
+    Returns:
+        The fully-formed GetFeature request URL.
+    """
+    is_wfs2 = version.startswith("2")
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py:469]: Done in f2a34b1 for the second nit — `add_wfs` now records `metadata.featureCount` (= len of the fetched features), matching the UI. On the first nit: `version.startswith("2")` is intentional parity with the TypeScript `createWfsGetFeatureUrl`; keeping the two in lockstep is preferable to diverging the heuristic here.

File: python/tests/test_project.py [hunk 6]
```diff
@@ -81,6 +81,170 @@ def test_cog_layer_restore_shape():
...
+def test_vector_layer_tiles_mode():
+    layer = project.vector_layer("V", "https://e/data.fgb", render_mode="tiles")
+    assert layer["type"] == "vector-tiles"
+    assert layer["source"]["type"] == "vector"
+    assert layer["metadata"]["vectorState"]["renderMode"] == "tiles"
+
+
+def test_vector_layer_invalid_render_mode():
+    with pytest.raises(ValueError):
+        project.vector_layer("V", "https://e/x", render_mode="bogus")
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/tests/test_project.py]: Done in f2a34b1 — the new `pytest.raises(ValueError)` assertions in test_project.py now use `match=` (render_mode, tile_type, video URL/corner, non-https), so each asserts its specific failure rather than any ValueError.

File: python/src/geolibre/geolibre.py [hunk 7]
```diff
@@ -24,6 +24,56 @@
...
+        raise ImportError(
+            "Reading a local vector file requires GeoPandas. Install it with "
+            "`pip install geopandas`, or pass a URL to a hosted dataset instead."
+        ) from exc
+    # GeoPandas' GDAL-backed read_file may lack the Parquet driver depending on
+    # the GDAL build, so dispatch (Geo)Parquet to the dedicated reader.
+    if file_path.suffix.lower() in (".parquet", ".geoparquet", ".pq"):
+        gdf = geopandas.read_parquet(file_path)
+    else:
+        gdf = geopandas.read_file(file_path)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py:70]: Done in 9ae568a — `_read_local_vector` now takes a `data_format` hint that overrides suffix detection, and `add_vector` threads it through, so a local GeoParquet under a non-standard name uses `geopandas.read_parquet`.

File: python/src/geolibre/project.py [hunk 8]
```diff
@@ -304,6 +309,427 @@ def cog_layer(
...
+        "sourceLayers": source_layers,
+        "tileType": tile_type,
+    }
+    layer["metadata"] = {
+        "sourceKind": "pmtiles-url",
+        "externalNativeLayer": True,
+        "sourceId": source_id,
+        "tileType": tile_type,
+        "sourceLayers": source_layers,
+        "nativeLayerIds": [],
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py]: Done in 9ae568a — confirmed the bug (isExternalNativeLayer() gates on nativeLayerIds.length and "pmtiles" has no fallback in syncLayer). `pmtiles_layer` now sets nativeLayerIds to [`<id>-raster`] for raster and [`<id>`] for vector, matching the fallbacks ensurePMTilesExternalLayer computes.

File: python/src/geolibre/project.py [hunk 9]
```diff
@@ -304,6 +309,427 @@ def cog_layer(
...
+
+    Returns:
+        A layer dict for the project's ``layers`` array.
+
+    Raises:
+        ValueError: If ``urls`` is empty, any URL is not ``https://`` (the
+            browser's ``media-src`` CSP blocks ``http://``), or ``coordinates``
+            is not four ``[lng, lat]`` pairs.
+    """
+    clean_urls = [u for u in urls if isinstance(u, str) and u]
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py]: Done in 9ae568a — `video_layer` now raises `ValueError` on any None/non-string/empty entry instead of silently dropping it (added a test).

File: python/src/geolibre/project.py [hunk 10]
```diff
@@ -304,6 +309,427 @@ def cog_layer(
...
+        vector_state["ingestMode"] = ingest_mode
+    layer["metadata"] = {
+        "sourceKind": "maplibre-gl-vector",
+        "externalNativeLayer": True,
+        # The control owns its layers' paint; the core sync must not re-apply it.
+        "controlOwnsPaint": True,
+        "identifiable": False,
+        # The control re-derives these on load; the post-load sync overwrites
+        # them with the real ids it creates, so empty placeholders are fine.
+        "nativeLayerIds": [],
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py:550]: Done in 9ae568a — added a comment explaining the empty list is safe for the geojson path (restoreVectorLayers adopts via sourceKind+externalNativeLayer, then syncVectorLayersToStore fills in the real ids) and noting the render_mode="tiles" syncVectorTileLayer fall-through caveat.

File: python/src/geolibre/geolibre.py [hunk 11]
```diff
@@ -288,6 +338,367 @@ def add_cog(
...
+                    data,
+                    render_mode=render_mode,
+                    data_format=data_format,
+                    source_layer=source_layer,
+                    **style,
+                )
+            )
+        if hasattr(data, "__geo_interface__"):
+            return self.add_geojson(data, name=name, **style)
+        fc = _read_local_vector(data)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py]: Done in 9ae568a — `add_vector` now warns when render_mode/source_layer are passed for a local file (they only apply to remote URLs). Note data_format is no longer ignored for local files — it is now threaded into the GeoParquet reader per the related CodeRabbit fix.

File: python/src/geolibre/geolibre.py [hunk 12]
```diff
@@ -288,6 +338,367 @@ def add_cog(
...
+    def add_wfs(
+        self,
+        endpoint: str,
+        type_name: str,
+        name: str = "WFS Layer",
+        *,
+        version: str = "2.0.0",
+        output_format: str = "application/json",
+        srs_name: str = "EPSG:4326",
+        max_features: int | None = None,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py]: Done in d2a7bb9 — `add_wfs` now defaults `max_features` to 1000 (matching the UI's WFS dialog) since the GeoJSON response is inlined; documented that `None` requests every feature.

File: python/src/geolibre/project.py [hunk 13]
```diff
@@ -304,6 +309,427 @@ def cog_layer(
...
+    url: str,
+    *,
+    tile_size: int = 256,
+    **style: Any,
+) -> dict[str, Any]:
+    """Build a WMTS layer from a tile URL template.
+
+    Args:
+        name: Layer display name.
+        url: A WMTS tile URL template (``{z}/{y}/{x}``).
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py]: Done in d2a7bb9 — clarified the `wmts_layer` docstring that the template uses WMTS REST `{z}/{y}/{x}` order (row before column), explicitly contrasting it with the XYZ `{z}/{x}/{y}` order in `add_tile_layer`.

File: python/src/geolibre/project.py [hunk 14]
```diff
@@ -304,6 +309,448 @@ def cog_layer(
...
+
+    Args:
+        endpoint: Base service URL (may already carry a query string).
+        params: Ordered ``(key, value)`` pairs to append.
+
+    Returns:
+        The endpoint with the encoded query string appended.
+    """
+    if "?" in endpoint:
+        separator = "" if endpoint.endswith(("?", "&")) else "&"
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py]: Done in 00f8aa4 — `_append_query` now splits off any `#fragment` (via `str.partition`), appends the query to the base, and re-attaches the fragment, so a fragment endpoint no longer strips the params.

File: python/src/geolibre/geolibre.py [hunk 15]
```diff
@@ -288,6 +346,379 @@ def add_cog(
             )
         )
 
+    def add_raster(
+        self,
+        url: str,
+        name: str = "Raster",
+        *,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py:353]: Done in 00f8aa4 — the `__geo_interface__` path now warns when render_mode/data_format/source_layer are passed, consistent with the local-file path.

File: python/src/geolibre/project.py [hunk 16]
```diff
@@ -304,6 +309,448 @@ def cog_layer(
...
+        source_layer: A single source-layer name (convenience for the common
+            single-layer case).
+        **style: Style overrides merged into the default layer style.
+
+    Returns:
+        A layer dict for the project's ``layers`` array.
+    """
+    layer = _layer_base(name, "vector-tiles", **style)
+    source: dict[str, Any] = {"type": "vector", "url": url}
+    if source_layers:
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/project.py:585]: Done in 00f8aa4 — `vector_tiles_layer` now warns when both `source_layers` and `source_layer` are supplied instead of silently dropping the latter.

File: python/src/geolibre/geolibre.py [hunk 17]
```diff
@@ -288,6 +346,379 @@ def add_cog(
...
+        Returns:
+            The id of the added layer.
+        """
+        url = _project.wfs_getfeature_url(
+            endpoint,
+            type_name,
+            version=version,
+            output_format=output_format,
+            srs_name=srs_name,
+            max_features=max_features,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py:479]: Done in da300a5 — `add_wfs` now catches the generic "Could not interpret input as GeoJSON" (the XML case) and re-raises an actionable message naming the typeName/output_format; size and SSRF errors keep their original message.

File: python/src/geolibre/geolibre.py [hunk 18]
```diff
@@ -24,6 +25,63 @@
...
+    else:
+        gdf = geopandas.read_file(file_path)
+    if gdf.crs is not None:
+        gdf = gdf.to_crs(epsg=4326)
+    # Round-trip through GeoPandas' own GeoJSON writer so numpy/datetime property
+    # values become plain JSON the widget bus can serialize.
+    geojson = gdf.to_json()
+    # Cap the inlined payload like load_featurecollection does for URL/file
+    # GeoJSON; a format like Shapefile can expand sharply once converted.
+    if len(geojson.encode("utf-8")) > _project._MAX_GEOJSON_BYTES:
```
[the remark this answers is not in the document]
Comment (giswqs) [reply python/src/geolibre/geolibre.py:78]: Holding off on this one (low confidence, leaving open for a maintainer call). The post-conversion 50 MB cap already guards the memory that actually matters here — the inlined GeoJSON payload synced over the widget bus. A pre-read size check can't see the converted size (formats expand unpredictably), so it would only be a heuristic warning on the raw file that fires on legitimately-large local reads and adds noise without preventing the GeoPandas read. Happy to add an entry-point warn-hint if you'd prefer it.
